<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>PyPedal: Package PyPedal.pyp_metrics</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.7 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li id="current"><a href="namespaces.html"><span>Packages</span></a></li>
    <li><a href="classes.html"><span>Classes</span></a></li>
    <li><a href="files.html"><span>Files</span></a></li>
  </ul></div>
<div class="tabs">
  <ul>
    <li><a href="namespaces.html"><span>Packages</span></a></li>
    <li><a href="namespacemembers.html"><span>Package&nbsp;Functions</span></a></li>
  </ul></div>
<div class="nav">
<b>PyPedal</b>.<a class="el" href="namespacePyPedal_1_1pyp__metrics.html">pyp_metrics</a></div>
<h1>Package PyPedal.pyp_metrics</h1>
<p>
<table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#f92db7d70e697a1f299c1b01775e9378">min_max_f</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#f92db7d70e697a1f299c1b01775e9378">min_max_f()</a> takes a pedigree and returns a list of the individuals with the n largest and n smallest coefficients of inbreeding.  <a href="#f92db7d70e697a1f299c1b01775e9378"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#3e7338b4892ff4cdedeb907a08114414">a_effective_founders_lacy</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#3e7338b4892ff4cdedeb907a08114414">a_effective_founders_lacy()</a> calculates the number of effective founders in a pedigree using the exact method of Lacy.  <a href="#3e7338b4892ff4cdedeb907a08114414"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#7fd43aa5db477016b9a1cceb7f2eb52d">effective_founders_lacy</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#7fd43aa5db477016b9a1cceb7f2eb52d">effective_founders_lacy()</a> calculates the number of effective founders in a pedigree using the exact method of Lacy.  <a href="#7fd43aa5db477016b9a1cceb7f2eb52d"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#2bd174a8cf151ec7b4ed7af7411858b0">a_effective_founders_boichard</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#2bd174a8cf151ec7b4ed7af7411858b0">a_effective_founders_boichard()</a> uses the algorithm in Appendix A of Boichard et al.  <a href="#2bd174a8cf151ec7b4ed7af7411858b0"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8821be8acfb9095c5f3776906568110b">a_effective_ancestors_definite</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8821be8acfb9095c5f3776906568110b">a_effective_ancestors_definite()</a> uses the algorithm in Appendix B of Boichard et al.  <a href="#8821be8acfb9095c5f3776906568110b"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#dcfac272aea7f8f5d0e43ef9d400f6fa">a_effective_ancestors_indefinite</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#dcfac272aea7f8f5d0e43ef9d400f6fa">a_effective_ancestors_indefinite()</a> uses the approach outlined on pages 9 and 10 of Boichard et al.  <a href="#dcfac272aea7f8f5d0e43ef9d400f6fa"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e298ad3dcbaf85dc2cbf946d188a5265">a_coefficients</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e298ad3dcbaf85dc2cbf946d188a5265">a_coefficients()</a> writes population average coefficients of inbreeding and relationship to a file, as well as individual animal IDs and coefficients of inbreeding.  <a href="#e298ad3dcbaf85dc2cbf946d188a5265"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#b054b0e4d468ae4013645af61102f6f3">fast_a_coefficients</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">a_fast_coefficients() writes population average coefficients of inbreeding and relationship to a file, as well as individual animal IDs and coefficients of inbreeding.  <a href="#b054b0e4d468ae4013645af61102f6f3"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#11dc9cb957ff03c709cecd2cbb996e33">theoretical_ne_from_metadata</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#11dc9cb957ff03c709cecd2cbb996e33">theoretical_ne_from_metadata()</a> computes the theoretical effective population size based on the number of sires and dams contained in a pedigree metadata object.  <a href="#11dc9cb957ff03c709cecd2cbb996e33"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#9839b790d3161aaf080c649adcdaff26">pedigree_completeness</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#9839b790d3161aaf080c649adcdaff26">pedigree_completeness()</a> computes the proportion of known ancestors in the pedigree of each animal in the population for a user-determined number of generations.  <a href="#9839b790d3161aaf080c649adcdaff26"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#43a82ce2cfd34b53ddf6d29c4671d480">common_ancestors</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#43a82ce2cfd34b53ddf6d29c4671d480">common_ancestors()</a> returns a list of the ancestors that two animals share in common.  <a href="#43a82ce2cfd34b53ddf6d29c4671d480"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#ec3083eaf96d141f21de2ee947cdeee8">related_animals</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#ec3083eaf96d141f21de2ee947cdeee8">related_animals()</a> returns a list of the ancestors of an animal.  <a href="#ec3083eaf96d141f21de2ee947cdeee8"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#84284c26af4c946ca9f49657b6d92fcf">relationship</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#84284c26af4c946ca9f49657b6d92fcf">relationship()</a> returns the coefficient of relationship for two animals, anim_a and anim_b.  <a href="#84284c26af4c946ca9f49657b6d92fcf"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#41c2baf2ca7fae116732adc32ff3a51b">mating_coi</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#41c2baf2ca7fae116732adc32ff3a51b">mating_coi()</a> returns the coefficient of inbreeding of offspring of a mating between two animals, anim_a and anim_b.  <a href="#41c2baf2ca7fae116732adc32ff3a51b"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e051130e25232f609ede648a4dea9986">mating_coi_group</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e051130e25232f609ede648a4dea9986">mating_coi_group()</a> returns the coefficients of inbreeding of offspring of a series of matings, as well as a list of minimum-inbreeding matings.  <a href="#e051130e25232f609ede648a4dea9986"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e86f472bfb3705f8a75323d161f6a358">effective_founder_genomes</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e86f472bfb3705f8a75323d161f6a358">effective_founder_genomes()</a> simulates the random segregation of founder alleles through a pedigree.  <a href="#e86f472bfb3705f8a75323d161f6a358"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#4fc7310e1c6484240ec0571a654cc64a">generation_intervals</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#4fc7310e1c6484240ec0571a654cc64a">generation_intervals()</a> computes the average age of parents at the time of birth of their first (oldest) offspring.  <a href="#4fc7310e1c6484240ec0571a654cc64a"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#553d55d2e917b00baf2f02a7cceacd70">generation_intervals_all</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#553d55d2e917b00baf2f02a7cceacd70">generation_intervals_all()</a> computes the average age of parents at the time of birth of their offspring.  <a href="#553d55d2e917b00baf2f02a7cceacd70"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8bb9c8d14721828c6d87bcd0d4e24658">founder_descendants</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8bb9c8d14721828c6d87bcd0d4e24658">founder_descendants()</a> returns a dictionary containing a list of descendants of each founder in the pedigree.  <a href="#8bb9c8d14721828c6d87bcd0d4e24658"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#132a4b6580f08cb10db3b376849a2bfe">descendants</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#132a4b6580f08cb10db3b376849a2bfe">descendants()</a> uses pedigree metadata to walk a pedigree and return a list of all of the descendants of a given animal.  <a href="#132a4b6580f08cb10db3b376849a2bfe"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e1cfcdb6871cd70b61e99da591975b58">dropped_ancestral_inbreeding</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e1cfcdb6871cd70b61e99da591975b58">dropped_ancestral_inbreeding()</a> uses a gene dropping approach to calculate ancestral inbreeding, the probability of an individual inheriting an allele that has undergone inbreeding in the past at least once.  <a href="#e1cfcdb6871cd70b61e99da591975b58"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">def&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#6185ce91b4815918ed935f1d4fb19714">ballou_ancestral_inbreeding</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight"><a class="el" href="namespacePyPedal_1_1pyp__metrics.html#6185ce91b4815918ed935f1d4fb19714">ballou_ancestral_inbreeding()</a> calculates ancestral inbreeding, the probability of an individual inheriting an allele that has undergone inbreeding in the past at least once, using the method of Ballou (1997).  <a href="#6185ce91b4815918ed935f1d4fb19714"></a><br></td></tr>
</table>
<hr><h2>Function Documentation</h2>
<a class="anchor" name="f92db7d70e697a1f299c1b01775e9378"></a><!-- doxytag: member="PyPedal::pyp_metrics::min_max_f" ref="f92db7d70e697a1f299c1b01775e9378" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.min_max_f           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>n</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>forma</em> = <code>'dense'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#f92db7d70e697a1f299c1b01775e9378">min_max_f()</a> takes a pedigree and returns a list of the individuals with the n largest and n smallest coefficients of inbreeding. 
<p>
Individuals with CoI of zero are not included. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>n</em>&nbsp;</td><td>An integer (optional, default is 10). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>forma</em>&nbsp;</td><td>If A must be formed should dense or sparse matrices be used? </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>Lists</em>&nbsp;</td><td>of the individuals with the n largest and the n smallest CoI in the pedigree as (ID, CoI) tuples.<div class="fragment"><pre class="fragment">Given a pedigree or relationship matrix, return a list of the
individuals with the n largest and n smallest coefficients of
inbreeding; individuals with CoI of zero are not included.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00063">63</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="3e7338b4892ff4cdedeb907a08114414"></a><!-- doxytag: member="PyPedal::pyp_metrics::a_effective_founders_lacy" ref="3e7338b4892ff4cdedeb907a08114414" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.a_effective_founders_lacy           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#3e7338b4892ff4cdedeb907a08114414">a_effective_founders_lacy()</a> calculates the number of effective founders in a pedigree using the exact method of Lacy. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary of results, including the effective founder number.<div class="fragment"><pre class="fragment">Calculate the number of effective founders in a pedigree using the exact method of Lacy.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00117">117</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="7fd43aa5db477016b9a1cceb7f2eb52d"></a><!-- doxytag: member="PyPedal::pyp_metrics::effective_founders_lacy" ref="7fd43aa5db477016b9a1cceb7f2eb52d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.effective_founders_lacy           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#7fd43aa5db477016b9a1cceb7f2eb52d">effective_founders_lacy()</a> calculates the number of effective founders in a pedigree using the exact method of Lacy. 
<p>
This version of the routine <a class="el" href="namespacePyPedal_1_1pyp__metrics.html#3e7338b4892ff4cdedeb907a08114414">a_effective_founders_lacy()</a> is designed to work with larger pedigrees as it forms "familywise" relationship matrices rather than a "populationwise" relationship matrix. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary of results, including the effective founder number.<div class="fragment"><pre class="fragment">Calculate the number of effective founders in a pedigree using the exact method of Lacy.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00216">216</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="2bd174a8cf151ec7b4ed7af7411858b0"></a><!-- doxytag: member="PyPedal::pyp_metrics::a_effective_founders_boichard" ref="2bd174a8cf151ec7b4ed7af7411858b0" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.a_effective_founders_boichard           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gen</em> = <code>''</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#2bd174a8cf151ec7b4ed7af7411858b0">a_effective_founders_boichard()</a> uses the algorithm in Appendix A of Boichard et al. 
<p>
(1996) to compute the effective founder number for myped. Note that results from this function will not necessarily match those from <a class="el" href="namespacePyPedal_1_1pyp__metrics.html#3e7338b4892ff4cdedeb907a08114414">a_effective_founders_lacy()</a>. Boichard's algorithm requires information about the GENERATION of animals. If you do not provide an input pedigree with generations things may not work. By default the most recent generation -- the generation with the largest generation ID -- will be used as the reference population. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gen</em>&nbsp;</td><td>Generation of interest. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>effective founder number.<div class="fragment"><pre class="fragment">The algorithm in Appendix A of Boichard et al. (1996) is not very well written.
a_effective_founders_boichard() implements that algorithm (successfully, I hope).
Note that answers from this function will not necessarily match those from
a_effective_founders_lacy().
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00328">328</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="8821be8acfb9095c5f3776906568110b"></a><!-- doxytag: member="PyPedal::pyp_metrics::a_effective_ancestors_definite" ref="8821be8acfb9095c5f3776906568110b" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.a_effective_ancestors_definite           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gen</em> = <code>''</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8821be8acfb9095c5f3776906568110b">a_effective_ancestors_definite()</a> uses the algorithm in Appendix B of Boichard et al. 
<p>
(1996) to compute the effective ancestor number for a myped pedigree. NOTE: One problem here is that if you pass a pedigree WITHOUT generations and error is not thrown. You simply end up wth a list of generations that contains the default value for Animal() objects, 0. Boichard's algorithm requires information about the GENERATION of animals. If you do not provide an input pedigree with generations things may not work. By default the most recent generation -- the generation with the largest generation ID -- will be used as the reference population. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gen</em>&nbsp;</td><td>Generation of interest. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>effective ancestor number.<div class="fragment"><pre class="fragment">The algorithm in Appendix B of Boichard et al. (1996) is not very well written.
a_effective_ancestors_definite() implements that algorithm (successfully, I hope).

NOTE: One problem here is that if you pass a pedigree WITHOUT generations an error
is not thrown.  You simply end up wth a list of generations that contains the
default value for Animal() objects, 0.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00477">477</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="dcfac272aea7f8f5d0e43ef9d400f6fa"></a><!-- doxytag: member="PyPedal::pyp_metrics::a_effective_ancestors_indefinite" ref="dcfac272aea7f8f5d0e43ef9d400f6fa" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.a_effective_ancestors_indefinite           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gen</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>n</em> = <code>25</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#dcfac272aea7f8f5d0e43ef9d400f6fa">a_effective_ancestors_indefinite()</a> uses the approach outlined on pages 9 and 10 of Boichard et al. 
<p>
(1996) to compute approximate upper and lower bounds for f_a. This is much more tractable for large pedigrees than the exact computation provided in <a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8821be8acfb9095c5f3776906568110b">a_effective_ancestors_definite()</a>. NOTE: One problem here is that if you pass a pedigree WITHOUT generations an error is not thrown. You simply end up wth a list of generations that contains the default value for NewAnimal() objects, 0. NOTE: If you pass a value of n that is greater than the actual number of ancestors in the pedigree then strange things happen. As a stop-gap, <a class="el" href="namespacePyPedal_1_1pyp__metrics.html#dcfac272aea7f8f5d0e43ef9d400f6fa">a_effective_ancestors_indefinite()</a> will detect that case and replace n with the number of founders - 1. Boichard's algorithm requires information about the GENERATION of animals. If you do not provide an input pedigree with generations things may not work. By default the most recent generation -- the generation with the largest generation ID -- will be used as the reference population. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gen</em>&nbsp;</td><td>Generation of interest. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>n</em>&nbsp;</td><td>The number of rounds to iterate for solutions. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>effective ancestor number.<div class="fragment"><pre class="fragment">a_effective_ancestors_indefinite() uses the approach outlined on pages 9 and 10 of
Boichard et al. (1996) to compute approximate upper and lower bounds for f_a.  This
is much more tractable for large pedigrees than the exact computation provided in
a_effective_ancestors_definite().
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00745">745</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="e298ad3dcbaf85dc2cbf946d188a5265"></a><!-- doxytag: member="PyPedal::pyp_metrics::a_coefficients" ref="e298ad3dcbaf85dc2cbf946d188a5265" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.a_coefficients           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>method</em> = <code>'nrm'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e298ad3dcbaf85dc2cbf946d188a5265">a_coefficients()</a> writes population average coefficients of inbreeding and relationship to a file, as well as individual animal IDs and coefficients of inbreeding. 
<p>
Some pedigrees are too large for fast_a_matrix() or fast_a_matrix_r() -- an array that large cannot be allocated due to memory restrictions -- and will result in a value of -999.9 for all outputs. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>method</em>&nbsp;</td><td>If no relationship matrix is passed, determines which procedure should be called to build one (nrm|frm). </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary of non-zero individual inbreeding coefficients.<div class="fragment"><pre class="fragment">Write population average coefficients of inbreeding and relationship to a
file, as well as individual animal IDs and coefficients of inbreeding.  Some
pedigrees are too large for fast_a_matrix() or fast_a_matrix_r()
-- an array that large cannot be allocated due to memory restrictions -- and will
result in a value of -999.9 for all outputs.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l00997">997</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="b054b0e4d468ae4013645af61102f6f3"></a><!-- doxytag: member="PyPedal::pyp_metrics::fast_a_coefficients" ref="b054b0e4d468ae4013645af61102f6f3" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.fast_a_coefficients           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>a</em> = <code>''</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>method</em> = <code>'nrm'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>debug</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>storage</em> = <code>'dense'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
a_fast_coefficients() writes population average coefficients of inbreeding and relationship to a file, as well as individual animal IDs and coefficients of inbreeding. 
<p>
It returns a list of non-zero individual CoI. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>a</em>&nbsp;</td><td>A numerator relationship matrix (optional). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>method</em>&nbsp;</td><td>If no relationship matrix is passed, determines which procedure should be called to build one (nrm|frm). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>debug</em>&nbsp;</td><td>Print deubgging messages if 1, don't print otherwise. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>storage</em>&nbsp;</td><td>Use dense or sparse matrix storage. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary of non-zero individual inbreeding coefficients.<div class="fragment"><pre class="fragment">a_fast_coefficients() writes population average coefficients of inbreeding and
relationship to a file, as well as individual animal IDs and coefficients of
inbreeding.  It returns a list of non-zero individual CoI.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01124">1124</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="11dc9cb957ff03c709cecd2cbb996e33"></a><!-- doxytag: member="PyPedal::pyp_metrics::theoretical_ne_from_metadata" ref="11dc9cb957ff03c709cecd2cbb996e33" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.theoretical_ne_from_metadata           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#11dc9cb957ff03c709cecd2cbb996e33">theoretical_ne_from_metadata()</a> computes the theoretical effective population size based on the number of sires and dams contained in a pedigree metadata object. 
<p>
Writes results to an output file. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>True</em>&nbsp;</td><td>(1) on success, false (0) on failure<div class="fragment"><pre class="fragment">theoretical_ne_from_metadata() computes the theoretical effective population
size based on the number of sires and dams contained in a pedigree metadata
object.  Writes results to an output file.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01249">1249</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="9839b790d3161aaf080c649adcdaff26"></a><!-- doxytag: member="PyPedal::pyp_metrics::pedigree_completeness" ref="9839b790d3161aaf080c649adcdaff26" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.pedigree_completeness           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gens</em> = <code>4</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#9839b790d3161aaf080c649adcdaff26">pedigree_completeness()</a> computes the proportion of known ancestors in the pedigree of each animal in the population for a user-determined number of generations. 
<p>
Also, the mean pedcomps for all animals and for all animals that are not founders are computed as summary statistics. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gens</em>&nbsp;</td><td>The number of generations the pedigree should be traced for completeness. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>Dictionary</em>&nbsp;</td><td>of summary statistics<div class="fragment"><pre class="fragment">pedigree_completeness() computes the proportion of known ancestors in the
pedigree of each animal in the population for a user-determined number of
generations. Also, the mean pedcomps for all animals and for all animals
that are not founders are computed as summary statistics.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01290">1290</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="43a82ce2cfd34b53ddf6d29c4671d480"></a><!-- doxytag: member="PyPedal::pyp_metrics::common_ancestors" ref="43a82ce2cfd34b53ddf6d29c4671d480" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.common_ancestors           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim_a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim_b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#43a82ce2cfd34b53ddf6d29c4671d480">common_ancestors()</a> returns a list of the ancestors that two animals share in common. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>anim_a</em>&nbsp;</td><td>The renumbered ID of the first animal, a. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>anim_b</em>&nbsp;</td><td>The renumbered ID of the second animal, b. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>list of animals related to anim_a AND anim_b<div class="fragment"><pre class="fragment">common_ancestors() returns a list of the ancestors that two animals share in common.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01414">1414</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="ec3083eaf96d141f21de2ee947cdeee8"></a><!-- doxytag: member="PyPedal::pyp_metrics::related_animals" ref="ec3083eaf96d141f21de2ee947cdeee8" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.related_animals           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#ec3083eaf96d141f21de2ee947cdeee8">related_animals()</a> returns a list of the ancestors of an animal. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>anim</em>&nbsp;</td><td>The renumbered ID of an animal, a. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>list of animals related to anim_a<div class="fragment"><pre class="fragment">related_animals() returns a list of the ancestors of an animal.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01446">1446</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="84284c26af4c946ca9f49657b6d92fcf"></a><!-- doxytag: member="PyPedal::pyp_metrics::relationship" ref="84284c26af4c946ca9f49657b6d92fcf" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.relationship           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim_a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim_b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>renumber</em> = <code>False</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#84284c26af4c946ca9f49657b6d92fcf">relationship()</a> returns the coefficient of relationship for two animals, anim_a and anim_b. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>anim_a</em>&nbsp;</td><td>The renumbered ID of an animal, a. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>anim_b</em>&nbsp;</td><td>The renumbered ID of an animal, b. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>renumber</em>&nbsp;</td><td>Renumber the pedigree if it has not yet been renumbered. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>coefficient of relationship of anim_a and anim_b<div class="fragment"><pre class="fragment">relationship() returns the coefficient of relationship for two
animals, anim_a and anim_b.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01471">1471</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="41c2baf2ca7fae116732adc32ff3a51b"></a><!-- doxytag: member="PyPedal::pyp_metrics::mating_coi" ref="41c2baf2ca7fae116732adc32ff3a51b" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.mating_coi           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim_a</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anim_b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gens</em> = <code>0</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#41c2baf2ca7fae116732adc32ff3a51b">mating_coi()</a> returns the coefficient of inbreeding of offspring of a mating between two animals, anim_a and anim_b. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>anim_a</em>&nbsp;</td><td>The renumbered ID of an animal, a. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>anim_b</em>&nbsp;</td><td>The renumbered ID of an animal, b. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gens</em>&nbsp;</td><td>The number of generations from the pedigree to be used for calculating CoI. By default, gens=0. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>coefficient of inbreeding of the offpsring of anim_a and anim_b<div class="fragment"><pre class="fragment">mating_coi() returns the coefficient of inbreeding of offspring of a
mating between two animals, anim_a and anim_b.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01557">1557</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="e051130e25232f609ede648a4dea9986"></a><!-- doxytag: member="PyPedal::pyp_metrics::mating_coi_group" ref="e051130e25232f609ede648a4dea9986" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.mating_coi_group           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>matings</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>names</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>gens</em> = <code>0</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e051130e25232f609ede648a4dea9986">mating_coi_group()</a> returns the coefficients of inbreeding of offspring of a series of matings, as well as a list of minimum-inbreeding matings. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>matings</em>&nbsp;</td><td>A list of proposed matings </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>names</em>&nbsp;</td><td>Indicates if the identifiers in 'matings' are names or animalIDs </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gens</em>&nbsp;</td><td>The number of generations from the pedigree to be used for calculating CoI. By default, gens=0. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>coefficient of inbreeding of the offpsring of anim_a and anim_b </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01690">1690</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="e86f472bfb3705f8a75323d161f6a358"></a><!-- doxytag: member="PyPedal::pyp_metrics::effective_founder_genomes" ref="e86f472bfb3705f8a75323d161f6a358" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.effective_founder_genomes           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>rounds</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>chrometype</em> = <code>'autosome'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>heterogametic</em> = <code>'m'</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>quiet</em> = <code>False</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e86f472bfb3705f8a75323d161f6a358">effective_founder_genomes()</a> simulates the random segregation of founder alleles through a pedigree. 
<p>
At present only two alleles are simulated for each founder. Summary statistics are computed on the most recent generation. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>rounds</em>&nbsp;</td><td>The number of times to simulate segregation through the entire pedigree. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>chrometype</em>&nbsp;</td><td>The type of chromosome to simulate ('autosome'|'sex'). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>heterogametic</em>&nbsp;</td><td>The heterogametic sex in your species ('f'|'m'). </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>quiet</em>&nbsp;</td><td>If False output is shown; if True output is not shown. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>The</em>&nbsp;</td><td>effective number of founder genomes based on 'rounds' of gene-drop simulations.<div class="fragment"><pre class="fragment">effective_founder_genomes() simulates the random segregation of founder alleles
through a pedigree.  At present only two alleles are simulated for each founder.
Summary statistics are computed on the most recent generation.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01778">1778</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="4fc7310e1c6484240ec0571a654cc64a"></a><!-- doxytag: member="PyPedal::pyp_metrics::generation_intervals" ref="4fc7310e1c6484240ec0571a654cc64a" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.generation_intervals           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>units</em> = <code>'y'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#4fc7310e1c6484240ec0571a654cc64a">generation_intervals()</a> computes the average age of parents at the time of birth of their first (oldest) offspring. 
<p>
This is implies that selection decisions are made at the time of birth of the first offspring. Average ages are computed for each of four paths: sire-son, sire-daughter, dam-son, and dam-daughter. An overall mean is computed, as well. IT IS IMPORTANT to note that if you DO NOT provide birthyears in your pedigree file that the returned dictionary will contain only zeroes! This is because when no birthyear is provided a default value (1900) is assigned to all animals in the pedigree. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>units</em>&nbsp;</td><td>A character indicating the units in which the generation lengths should be returned. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary containing the five average ages.<div class="fragment"><pre class="fragment">generation_intervals() computes the average age of parents at the time of
birth of their first (oldest) offspring.  This is implies that selection
decisions are made at the time of birth of the first offspring.  Average
ages are computed for each of four paths: sire-son, sire-daughter, dam-son,
and dam-daughter.  An overall mean is computed, as well.  IT IS IMPORTANT
to note that if you DO NOT provide birthyears in your pedigree file that the
returned dictionary will contain only zeroes!  This is because when no birthyear
is provided a default value (1900) is assigned to all animals in the pedigree.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l01980">1980</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="553d55d2e917b00baf2f02a7cceacd70"></a><!-- doxytag: member="PyPedal::pyp_metrics::generation_intervals_all" ref="553d55d2e917b00baf2f02a7cceacd70" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.generation_intervals_all           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>units</em> = <code>'y'</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#553d55d2e917b00baf2f02a7cceacd70">generation_intervals_all()</a> computes the average age of parents at the time of birth of their offspring. 
<p>
The computation is made using birth years for all known offspring of sires and dams, which implies discrete generations. Average ages are computed for each of four paths: sire-son, sire-daughter, dam-son, and dam-daughter. An overall mean is computed, as well. IT IS IMPORTANT to note that if you DO NOT provide birthyears in your pedigree file that the returned dictionary will contain only zeroes! This is because when no birthyear is provided a default value (1900) is assigned to all animals in the pedigree. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>units</em>&nbsp;</td><td>A character indicating the units in which the generation lengths should be returned. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary containing the five average ages.<div class="fragment"><pre class="fragment">generation_intervals_all() computes the average age of parents at the time of
birth of their offspring.  The computation is made using birth years for all
known offspring of sires and dams, which implies discrete generations.  Average
ages are computed for each of four paths: sire-son, sire-daughter, dam-son, and
dam-daughter.  An overall mean is computed, as well. IT IS IMPORTANT to note that
if you DO NOT provide birthyears in your pedigree file that the returned dictionary
will contain only zeroes!  This is because when no birthyear is provided a default
value (1900) is assigned to all animals in the pedigree.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l02173">2173</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="8bb9c8d14721828c6d87bcd0d4e24658"></a><!-- doxytag: member="PyPedal::pyp_metrics::founder_descendants" ref="8bb9c8d14721828c6d87bcd0d4e24658" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.founder_descendants           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#8bb9c8d14721828c6d87bcd0d4e24658">founder_descendants()</a> returns a dictionary containing a list of descendants of each founder in the pedigree. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>An instance of a PyPedal NewPedigree object. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary containing a list of descendants of each founder in the pedigree.<div class="fragment"><pre class="fragment">founder_descendants() returns a dictionary containing a list of descendants of
each founder in the pedigree.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l02315">2315</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="132a4b6580f08cb10db3b376849a2bfe"></a><!-- doxytag: member="PyPedal::pyp_metrics::descendants" ref="132a4b6580f08cb10db3b376849a2bfe" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.descendants           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>anid</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>_desc</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#132a4b6580f08cb10db3b376849a2bfe">descendants()</a> uses pedigree metadata to walk a pedigree and return a list of all of the descendants of a given animal. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>anid</em>&nbsp;</td><td>An animal ID </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A Python list of PyPedal Animal() objects. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>_desc</em>&nbsp;</td><td>A Python dictionary of descendants of animal anid. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>list of descendants of anid.<div class="fragment"><pre class="fragment">descendants() uses pedigree metadata to walk a pedigree and return a list of all
of the descendants of a given animal.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l02337">2337</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="e1cfcdb6871cd70b61e99da591975b58"></a><!-- doxytag: member="PyPedal::pyp_metrics::dropped_ancestral_inbreeding" ref="e1cfcdb6871cd70b61e99da591975b58" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.dropped_ancestral_inbreeding           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>rounds</em> = <code>100</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>loci</em> = <code>100</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>frequency</em> = <code>0.05</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>seed</em> = <code>5048665</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#e1cfcdb6871cd70b61e99da591975b58">dropped_ancestral_inbreeding()</a> uses a gene dropping approach to calculate ancestral inbreeding, the probability of an individual inheriting an allele that has undergone inbreeding in the past at least once. 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>rounds</em>&nbsp;</td><td>The number of times to simulate segregation through the entire pedigree. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>loci</em>&nbsp;</td><td>The number biallelic, unlinked loci to simulate. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>frequency</em>&nbsp;</td><td>The minor allele frequency. </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>seed</em>&nbsp;</td><td>The seed for the RNG. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary of ancestral inbreeding coefficients keyed to animal IDs.<div class="fragment"><pre class="fragment">dropped_ancestral_inbreeding() uses a gene dropping approach to calculate
ancestral inbreeding, the probability of an individual inheriting an allele
that has undergone inbreeding in the past at least once.
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l02384">2384</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<a class="anchor" name="6185ce91b4815918ed935f1d4fb19714"></a><!-- doxytag: member="PyPedal::pyp_metrics::ballou_ancestral_inbreeding" ref="6185ce91b4815918ed935f1d4fb19714" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def PyPedal.pyp_metrics.ballou_ancestral_inbreeding           </td>
          <td>(</td>
          <td class="paramtype">&nbsp;</td>
          <td class="paramname"> <em>pedobj</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
<a class="el" href="namespacePyPedal_1_1pyp__metrics.html#6185ce91b4815918ed935f1d4fb19714">ballou_ancestral_inbreeding()</a> calculates ancestral inbreeding, the probability of an individual inheriting an allele that has undergone inbreeding in the past at least once, using the method of Ballou (1997). 
<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>pedobj</em>&nbsp;</td><td>A PyPedal pedigree object. </td></tr>
  </table>
</dl>
<dl compact><dt><b>Return values:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>A</em>&nbsp;</td><td>dictionary of ancestral inbreeding coefficients keyed to animal IDs.<div class="fragment"><pre class="fragment">ballou_ancestral_inbreeding() calculates ancestral inbreeding,
the probability of an individual inheriting an allele that has
undergone inbreeding in the past at least once, using the method
of Ballou (1997).
</pre></div> </td></tr>
  </table>
</dl>

<p>
Definition at line <a class="el" href="pyp__metrics_8py-source.html#l02509">2509</a> of file <a class="el" href="pyp__metrics_8py-source.html">pyp_metrics.py</a>.
</div>
</div><p>
<hr size="1"><address style="align: right;"><small>Generated on Tue Sep 28 14:09:57 2010 for PyPedal by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.7 </small></address>
</body>
</html>
